bundles/org.openhab.binding.bluetooth.bluez/src/main/java/org/openhab/binding/bluetooth/bluez/internal/BlueZBridgeHandler.java

```diff
@@ -0,0 +1,234 @@
+        super.initialize();
+
+        // Load configuration
+        final BlueZAdapterConfiguration configuration = getConfigAs(BlueZAdapterConfiguration.class);
+        String addr = configuration.address;
+        if (addr != null) {
+            this.adapterAddress = new BluetoothAddress(addr.toUpperCase());
+        } else {
+            // If configuration does not contain adapter address to use, exit with error.
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "address not set");
+            return;
+        }
+
+        logger.debug("Creating BlueZ adapter with address '{}'", adapterAddress);
+        updateStatus(ThingStatus.UNKNOWN, ThingStatusDetail.CONFIGURATION_PENDING, "Initializing");
+        deviceManagerFactory.getPropertiesChangedHandler().addListener(this);
+        discoveryJob = scheduler.scheduleWithFixedDelay(this::initializeAndRefreshDevices, 5, 10, TimeUnit.SECONDS);
+    }
+
+    @Override
+    public void dispose() {
+        deviceManagerFactory.getPropertiesChangedHandler().removeListener(this);
+        logger.debug("Termination of DBus BlueZ handler");
+
+        Future<?> job = discoveryJob;
+        if (job != null) {
+            job.cancel(false);
+            discoveryJob = null;
+        }
+
+        BluetoothAdapter localAdatper = this.adapter;
+        if (localAdatper != null) {
+            localAdatper.stopDiscovery();
+            this.adapter = null;
+        }
+
+        super.dispose();
+    }
+
+    private @Nullable BluetoothAdapter prepareAdapter(DeviceManagerWrapper deviceManager) {
+        // next lets check if we can find our adapter in the manager.
+        BluetoothAdapter localAdapter = adapter;
+        if (localAdapter == null) {
+            BluetoothAddress localAddress = adapterAddress;
+            if (localAddress != null) {
+                localAdapter = adapter = deviceManager.getAdapter(localAddress);
+            } else {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "No adapter address provided");
+                return null;
+            }
+        }
+        if (localAdapter == null) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "Native adapter could not be found for address '" + adapterAddress + "'");
+            return null;
+        }
+        // now lets confirm that the adapter is powered
+        if (!localAdapter.isPowered()) {
+            localAdapter.setPowered(true);
+            // give the device some time to power on
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.NONE,
+                    "Adapter is not powered, attempting to turn on...");
+            return null;
+        }
+
+        // now lets make sure that discovery is turned on
+        if (!localAdapter.startDiscovery()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.NONE, "Starting discovery");
```

Comment:
I'm still not clear why it is OFFLINE here.
The message "starting discovery" seems to suggest that we are still initializing, so I would simply stay in UNKNOWN status that moment (until we know if the outcome is successful or not).
If I misunderstand the message and the comment above ("make sure that discovery is turned on") is actually correct by saying that we have discovered a situation where we expect discovery being enabled, while it isn't and we consider this a "permanent" issue with the adapter, then OFFLINE is indeed the right choice.

Could you help me understand which one is right here?

Reply:
Sometimes bluez will turn off discovery as part of it's own error recovery system when it encounters strange bluetooth situations it doesn't know how to handle. (i.e. an advertisement packet that has incorrect length values, or a low level hardware error). Bluez won't turn discovery back on automatically so we just make sure we keep it on by starting it if it is turned off.

Now turning on discovery might now work if the underlying hardware is not ready, in which case trying again later might prove effective, but there might be something wrong with the hardware as well in which case the OFFLINE seems appropriate. Since I can't tell whether a discovery failure is going to be temporary or permanent I think it would be safer just make the bridge OFFLINE until we can confirm that discovery is properly active.

WDYT?

Reply:
@cpmeister is right that BlueZ for some reason stops discovering sometimes. sometimes, restarting discovery is not even sufficient which is the reason why I coded some adapter recycling. INITIALIZING would make sense if this piece of code would only be called upon binding startup, but if we look closely, the prepareAdapter method is also called on each "refreshdevice" iteration.

We could as such observe situation ONLINE->INITIALIZING->ONLINE. Otherwise it is ONLINE->OFFLINE->ONLINE

In any case, I'm fine with both solutions, as I don't think there is an impact ?

Reply:
Ok, thanks for your explanation! So how about adapting the message to:
```suggestion
            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.NONE, "Trying to start discovery");
```

Reply:
done